File: haskell/Vec3.hs
-- You do not need to modify this module.
module Vec3 (Vec3(..), scale, norm, normalise, cross, dot) where

import Control.DeepSeq

data Vec3 = Vec3 {vecX :: Float, vecY :: Float, vecZ :: Float}
  deriving (Eq, Ord, Show, Read)

instance NFData Vec3 where
  rnf (Vec3 x y z) = x `deepseq` y `deepseq` z `deepseq` ()

instance Num Vec3 where
  Vec3 x1 y1 z1 + Vec3 x2 y2 z2 = Vec3 (x1+x2) (y1+y2) (z1+z2)
  Vec3 x1 y1 z1 - Vec3 x2 y2 z2 = Vec3 (x1-x2) (y1-y2) (z1-z2)
  Vec3 x1 y1 z1 * Vec3 x2 y2 z2 = Vec3 (x1*x2) (y1*y2) (z1*z2)
  negate (Vec3 x y z) = Vec3 (negate x) (negate y) (negate z)
  abs (Vec3 x y z) = Vec3 (abs x) (abs y) (abs z)
  signum (Vec3 x y z) = Vec3 (signum x) (signum y) (signum z)
  fromInteger a = Vec3 (fromInteger a) (fromInteger a) (fromInteger a)

instance Fractional Vec3 where
  recip (Vec3 x y z) = Vec3 (recip x) (recip y) (recip z)
  fromRational a = Vec3 (fromRational a) (fromRational a) (fromRational a)

scale :: Float -> Vec3 -> Vec3
scale a (Vec3 x y z) = Vec3 (a*x) (a*y) (a*z)

dot :: Vec3 -> Vec3 -> Float
dot (Vec3 x1 y1 z1) (Vec3 x2 y2 z2) =
  x1*x2 + y1*y2 + z1*z2

norm :: Vec3 -> Float
norm v = sqrt (dot v v)

normalise :: Vec3 -> Vec3
normalise v = (1 / norm v) `scale` v

cross :: Vec3 -> Vec3 -> Vec3
cross (Vec3 x1 y1 z1) (Vec3 x2 y2 z2) =
  Vec3 (y1*z2-z1*y2) (z1*x2-x1*z2) (x1*y2-y1*x2)

File: scala/src/main/scala/raytracer/Vec3.scala
package raytracer

final case class Vec3(x: Double, y: Double, z: Double) {

  @inline def +(that: Vec3) = Vec3(x + that.x, y + that.y, z + that.z)
  @inline def -(that: Vec3) = Vec3(x - that.x, y - that.y, z - that.z)
  @inline def *(that: Vec3) = Vec3(x * that.x, y * that.y, z * that.z)

  @inline def scale(a: Double): Vec3 = Vec3(a * x, a * y, a * z)
  @inline def dot(that: Vec3): Double =
    x * that.x + y * that.y + z * that.z
  @inline def norm: Double = math.sqrt(this dot this).toDouble
  @inline def normalise: Vec3 = scale (1.0f / norm)
  @inline def cross(that: Vec3): Vec3 =
    Vec3(y * that.z - z * that.y, z * that.x - x * that.z, x * that.y - y * that.x)

}

object Vec3 {
  val one = Vec3(1, 1, 1)
}
